src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java

```diff
@@ -13,12 +34,368 @@
+                boolean ableToDeliver = wareHouse.getTileEntity().checkInWareHouse(buildingToDeliver);
+                if(!ableToDeliver)
+                {
+                    return START_WORKING;
+                }
+                itemsToDeliver = new ArrayList<>(buildingToDeliver.getNeededItems());
+                return GATHER_IN_WAREHOUSE;
+            }
+        }
+        return START_WORKING;
+    }
+
+    private AIState gatherItemsFromWareHouse()
+    {
+        AbstractBuildingWorker ownBuilding = getOwnBuilding();
+        if (ownBuilding instanceof BuildingDeliveryman)
+        {
+            AbstractBuilding buildingToDeliver = ((BuildingDeliveryman) ownBuilding).getBuildingToDeliver();
+            if (buildingToDeliver != null)
+            {
+                if(itemsToDeliver.isEmpty() && hasTools(buildingToDeliver))
+                {
+                    return DELIVERY;
+                }
+
+                gatherItems(buildingToDeliver);
+                return GATHER_IN_WAREHOUSE;
+            }
+        }
+        return START_WORKING;
+    }
+
+    /**
+     * Check if the deliveryman has all the tools to make the delivery.
+     * @param buildingToDeliver the building to deliver to.
+     * @return true if is ready to deliver.
+     */
+    private boolean hasTools(final AbstractBuilding buildingToDeliver)
+    {
+        String requiredTool = buildingToDeliver.getRequiredTool();
+        if(requiredTool.isEmpty())
+        {
+            return true;
+        }
+
+        if(requiredTool.equals(Utils.PICKAXE))
+        {
+            return InventoryUtils.isPickaxeInTileEntity(worker.getInventoryCitizen(), buildingToDeliver.getNeededPickaxeLevel());
+        }
+        return InventoryUtils.isToolInTileEntity(worker.getInventoryCitizen(), requiredTool);
+    }
+
+    /**
+     * Gather item
+     * @param buildingToDeliver
+     */
+    private void gatherItems(final AbstractBuilding buildingToDeliver)
+    {
+        BlockPos position;
+        if(itemsToDeliver.isEmpty())
+        {
+            String tool = buildingToDeliver.getRequiredTool();
+            position = wareHouse.getTileEntity().getPositionOfChestWithTool(tool, Utils.PICKAXE.equals(tool) ? buildingToDeliver.getNeededPickaxeLevel() : -1);
+        }
+        else
+        {
+            //todo take at least the amount of items he needs.
+            ItemStack stack = itemsToDeliver.get(0);
+            position = wareHouse.getTileEntity().getPositionOfChestWithItemStack(stack);
+        }
+
+        if(position == null)
+        {
+            return;
+        }
+
+        if(!worker.isWorkerAtSiteWithMove(position, MIN_DISTANCE_TO_WAREHOUSE))
+        {
+            return;
+        }
+
+        TileEntity tileEntity = world.getTileEntity(position);
+        if(tileEntity instanceof TileEntityChest)
+        {
+            if (itemsToDeliver.isEmpty())
+            {
+                String tool = buildingToDeliver.getRequiredTool();
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'tool' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)